### PR TITLE
Add support for anonymous tasks

### PR DESCRIPTION
**Pull request by naormalca**
# Description
This PR continues the work that [goatsthatcode](https://github.com/goatsthatcode) did on https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1052, due to the fact that the author hasn't responded anymore, we are interested in resolving this issue.

This changes the way spans are initialized for the celery instrumentation hooks and handles the case where Celery tasks are not available in the registry and called anonymously using the signature() or send_task()

Fixes https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1002, https://github.com/open-telemetry/opentelemetry-python-contrib/issues/609
possibly addresses https://github.com/open-telemetry/opentelemetry-python-contrib/issues/784

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Ran changes against reproduced issue in repo linked in issue -> https://github.com/goatsthatcode/opentelemetry-instrumentation-celery-example
- [x] Added Unit Test to cover new conditional branch
- [x] Passed existing test suite

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Comment by srikanthccv** [comment]
The last time I tried running the change, it didn't fix the issue. Is it still the case?

**Comment by naormalca** [comment]
> The last time I tried running the change, it didn't fix the issue. Is it still the case?

the change is fixing the issue, it tested with a unit test and integration tests on my project. what scenario reproduced the issue for you?

File: instrumentation/opentelemetry-instrumentation-celery/src/opentelemetry/instrumentation/celery/__init__.py [hunk 1]
```diff
@@ -183,10 +183,17 @@ def _trace_before_publish(self, *args, **kwargs):
         task = utils.retrieve_task_from_sender(kwargs)
         task_id = utils.retrieve_task_id_from_message(kwargs)
 
-        if task is None or task_id is None:
+        if task_id is None:
             return
 
-        operation_name = f"{_TASK_APPLY_ASYNC}/{task.name}"
+        if task is None:
```
Comment (srikanthccv) [inline instrumentation/opentelemetry-instrumentation-celery/src/opentelemetry/instrumentation/celery/__init__.py:189]: A line below attaches the span to the task, which is retrieved later; what happens to it now?

Comment (ocelotl) [reply instrumentation/opentelemetry-instrumentation-celery/src/opentelemetry/instrumentation/celery/__init__.py:189]: You mean that [here](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1407/files#diff-ed0dafdf74eda90912242a368cd05845c26ce3178f34c4fea9816f3ff1e49307R211) `task` could be `None` with these changes?

Comment (russlevy) [reply instrumentation/opentelemetry-instrumentation-celery/src/opentelemetry/instrumentation/celery/__init__.py:189]: `task` isn't used later, all instances of `task` in this function only referenced `task.name`, and that has been replaced with `task_name`.

File: CHANGELOG.md [hunk 2]
```diff
@@ -134,6 +134,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   ([#1333](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1333))
 - `opentelemetry-instrumentation-asgi` Make ASGIGetter.get() compare all keys in a case insensitive manner.
   ([#1333](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1333))
+- `opentelemetry-instrumentation-celery` Add support for anonymous tasks.
+  ([#1407](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1407)
```
Comment (srikanthccv) [inline CHANGELOG.md:138]: Move this to unreleased section

**Review by srikanthccv — APPROVED** [review]

**Pull request merged**